include/modules/meta/base.hpp

```diff
@@ -115,39 +123,145 @@ namespace modules {
   // }}}
   // class definition : module {{{
 
+  /**
+   * \brief Base classes of every modules.
+   * \details
+   * To implement a simple module, the following methods must be implemented:
+   *   - #start(): virtual override
+   *   - #build(builder*, const string&) const: CRTP implementation
+   *
+   * Optionally, the following methods might be reimplemented:
+   *   - all the method of module_interface excepts #contents(), #halt(string)
+   *   - #wakeup(): CRTP implementation
+   *   - #teardown(): CRTP implementation
+   *   - #get_output(): CRTP implementation
+   *   - #get_format(): CRTP implementation
+   *   - #idle() : CRTP implementation
+   *
+   * \tparam Impl - The final type of the module
+   */
   template <class Impl>
   class module : public module_interface {
    public:
     module(const bar_settings bar, string name);
     ~module() noexcept;
 
-    string name() const;
-    bool running() const;
-    void stop();
-    void halt(string error_message);
-    void teardown();
-    string contents();
+    /**
+     * \brief Returns the name of the module.
+     * \details
+     * This method call doesn't need to be protected.
+     */
+    string name() const final;
+
+    /**
+     * \brief Returns true if the module is running
+     * \details
+     * This method call doesn't need to be protected
+     */
+    bool running() const final;
+
+    /**
+     * \brief Stops the module
+     * \details
+     * If you need to clean the internal state of your module, #teardown should be used.
+     *
+     * This method does nothing if #running() returns false.
+     * This method should call #wakeup() then #teardown() using CRTP.
+     * These calls MUST be protected by locking #m_modulelock.
+     */
+    void stop() override;
+
+    /**
+     * \brief Logs an error and stop the module
+     * \details
+     * Logs the given error and call #stop()
+     */
+    void halt(string error_message) final;
+
+    /**
+     * \brief Generates the output of the module
+     */
+    string contents() final;
 
    protected:
+    /**
+     * \brief Interrupts sleep
+     * \details
+     * This method is NOT protected
```

Comment:
It just to document that this method isn't protected and so you should not use data that are not protected. If I'm not wrong we can protect this method but I'm not sure this is useful, since the aim of this method is to wakeup the thread not to modify the class.

Reply:
Ah, so it means that the method itself doesn't lock the mutex?

Reply:
It means that when this method is called the mutex is indeed not locked.